File: clients/python/zeeplabs_orbit_client/files.py
from __future__ import annotations

from typing import Optional
from urllib.request import Request

from .types import FileResponse


class FilesClient:
    def __init__(self, client):
        self._client = client

    def _path(self, p: str) -> str:
        return f"files/{p.lstrip('/')}"

    def upload(self, filename: str, data: bytes, mime_type: str) -> FileResponse:
        import uuid
        boundary = uuid.uuid4().hex
        header = f"--{boundary}\r\nContent-Disposition: form-data; name=\"file\"; filename=\"{filename}\"\r\nContent-Type: {mime_type}\r\n\r\n"
        footer = f"\r\n--{boundary}--\r\n"
        body = header.encode() + data + footer.encode()
        content_type = f"multipart/form-data; boundary={boundary}"

        url = self._client._url(self._path(""))
        req = Request(
            url,
            data=body,
            headers={
                "Authorization": f"Bearer {self._client._config.jwt}",
                "Content-Type": content_type,
            },
            method="POST",
        )
        from urllib.request import urlopen
        with urlopen(req) as resp:
            import json
            result = json.loads(resp.read())
            return FileResponse(**result)

    def list(self, limit: int = 50, offset: int = 0) -> list[FileResponse]:
        data = self._client._request("GET", self._path(f"?limit={limit}&offset={offset}"))
        return [FileResponse(**f) for f in data]

    def get(self, id: str) -> FileResponse:
        data = self._client._request("GET", self._path(id))
        return FileResponse(**data)

    def delete(self, id: str) -> None:
        self._client._request("DELETE", self._path(id))

    def signed_url(self, id: str, ttl: int = 3600) -> str:
        data = self._client._request("GET", self._path(f"{id}/url?ttl={ttl}"))
        return data["url"]

File: clients/rust/src/files.rs
use crate::client::OrbitClient;
use crate::types::FileResponse;
use reqwest::multipart;

#[derive(Debug)]
pub struct FilesClient<'a> {
    client: &'a OrbitClient,
}

impl<'a> FilesClient<'a> {
    pub fn new(client: &'a OrbitClient) -> Self {
        Self { client }
    }

    fn path(&self, p: &str) -> String {
        format!("files/{}", p.trim_start_matches('/'))
    }

    pub async fn upload(
        &self,
        filename: &str,
        data: Vec<u8>,
        mime_type: &str,
    ) -> Result<FileResponse, reqwest::Error> {
        let part = multipart::Part::bytes(data)
            .file_name(filename.to_string())
            .mime_str(mime_type)
            .unwrap();
        let form = multipart::Form::new().part("file", part);

        let resp = self
            .client
            .http
            .post(self.client.url(&self.path("")))
            .multipart(form)
            .header(
                "Authorization",
                format!("Bearer {}", self.client.config.jwt),
            )
            .send()
            .await?;

        resp.json::<FileResponse>().await
    }

    pub async fn list(&self, limit: i64, offset: i64) -> Result<Vec<FileResponse>, reqwest::Error> {
        let resp = self
            .client
            .http
            .get(self.client.url(&self.path(&format!("?limit={}&offset={}", limit, offset))))
            .headers(self.client.headers())
            .send()
            .await?;
        resp.json::<Vec<FileResponse>>().await
    }

    pub async fn get(&self, id: &str) -> Result<FileResponse, reqwest::Error> {
        let resp = self
            .client
            .http
            .get(self.client.url(&self.path(id)))
            .headers(self.client.headers())
            .send()
            .await?;
        resp.json::<FileResponse>().await
    }

    pub async fn delete(&self, id: &str) -> Result<(), reqwest::Error> {
        self.client
            .http
            .delete(self.client.url(&self.path(id)))
            .headers(self.client.headers())
            .send()
            .await?;
        Ok(())
    }

    pub async fn signed_url(&self, id: &str, ttl: i64) -> Result<String, reqwest::Error> {
        #[derive(serde::Deserialize)]
        struct SignedUrlResp {
            url: String,
        }
        let resp = self
            .client
            .http
            .get(self.client.url(&self.path(&format!("{}/url?ttl={}", id, ttl))))
            .headers(self.client.headers())
            .send()
            .await?
            .json::<SignedUrlResp>()
            .await?;
        Ok(resp.url)
    }
}
